### PR TITLE
Optimize Known Methods and Known Versions

### PR DESCRIPTION
**Pull request by Anderman**
Optimize TakeStartLine function. TakeStartLine is now 2 times faster then the previous one

### CODE REVIEW

File: KestrelHttpServer.sln [hunk 1]
```diff
@@ -37,6 +37,8 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNet.Server.Kes
 EndProject
```
Comment (cesarbs) [inline KestrelHttpServer.sln]: Revert.

File: src/test3/Properties/launchSettings.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (cesarbs) [inline src/test3/Properties/launchSettings.json]: Revert.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs [hunk 3]
```diff
@@ -89,6 +89,43 @@ public int Take()
                 }
             }
         }
+        public void Skip(int bytesToSkip)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: nit: blank line before method

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs [hunk 4]
```diff
@@ -89,6 +89,43 @@ public int Take()
...
+        {
+            if (_block == null)
+            {
+                return;
+            }
+            var following = _block.End - _index;
+            if(following>=bytesToSkip)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: nit: space between `if` and `(`

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 5]
```diff
@@ -262,54 +275,47 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
-        /// The number 37 was chosen because that number allows for a perfect hash of the set of
-        /// "known strings" (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE, HTTP/1.0 and HTTP/1.1, where strings
-        /// with less than 8 characters have 0s appended to their ends to fill for the missing bytes).
-        /// </remarks>
-        /// <param name="begin">The iterator from which to start the known string lookup.</param>
-        /// <param name="end">The iterator pointing to the end of the input string.</param>
-        /// <param name="knownString">A reference to a pre-allocated known string, if the input matches any.</param>
-        /// <returns><c>true</c> if the input matches a known string, <c>false</c> otherwise.</returns>
-        public static bool GetKnownString(this MemoryPoolIterator2 begin, MemoryPoolIterator2 end, out string knownString)
+        public static bool getKnownMethod(ref MemoryPoolIterator2 scan, out string knownString)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: `getKnownMethod` -> `GetKnownMethod` (uppercase `G`)

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 6]
```diff
@@ -262,54 +275,47 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
+                    return true;
+                }
             }
-
-            var inputLong = begin.PeekLong();
-
-            if (inputLong == -1)
+            return false;
+        }
+        public static bool getKnownVersion(ref MemoryPoolIterator2 scan, out string knownString)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: uppercase G in method name

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs (hunk 5 above)
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: Please add the documentation back and adjust as necessary.

**Comment by cesarbs** [comment]
How did you measure that it's 2 times faster?

**Comment by Anderman** [comment]
I did the measurement  on the function `TakeStartLine` with an input string `GET / HTTP/1.1\r\n`. on Core CLR X64 Release version. At the moment it is 1.3 times faster. 
The previous test I didn't realize that the different versions has an big impact on the relative test 

For the test I created an version with both versions of TakeStartLine (Org and new). The test looked something like

```
public static void perf_TakeStartLine()
        {
            var socketInput = new SocketInput(new MemoryPool2(), null);
            var firstLine = Encoding.ASCII.GetBytes("GET / HTTP/1.1\r\n");

            var inputBuffer = socketInput.IncomingStart(128);
            Buffer.BlockCopy(firstLine, 0, inputBuffer.Data.Array, inputBuffer.Data.Offset, firstLine.Length);
            socketInput.IncomingComplete(firstLine.Length, null);

            var maxRequest = 10000000;
            var connectionContext = new ConnectionContext()
            {
                SocketInput = socketInput,
                ServerAddress = new ServerAddress(),
                DateHeaderValueManager = new DateHeaderValueManager(),
            };
            var frame = new Frame<object>(application: null, context: connectionContext);
            var sw = new Stopwatch();
            sw.Start();
            for (int i = 0; i < maxRequest; i++)
            {
                var success = frame.TakeStartLine_org(socketInput);
            }
            sw.Stop();

            var sw2 = new Stopwatch();
            sw2.Start();
            for (int i = 0; i < maxRequest; i++)
            {
                var success = frame.TakeStartLine(socketInput);
            }
            sw2.Stop();
            var org = (int)(((maxRequest)) / (sw.Elapsed.TotalMilliseconds));
            var improved = (int)(((maxRequest)) / (sw2.Elapsed.TotalMilliseconds));
            var factor = (1.0 * improved) / org;
            Console.WriteLine($"org:{org:D4} new:{improved:D4} req/ms {factor.ToString("0.0")}x ");
        }
```

File: test/Microsoft.AspNet.Server.KestrelTests/MemoryPoolIterator2Tests.cs [hunk 7]
```diff
@@ -195,21 +195,39 @@ public void PeekLongAtBlockBoundary(int blockBytes)
...
         [InlineData("GET/ HTTP/1.1", ' ', false, null)]
         [InlineData("get / HTTP/1.1", ' ', false, null)]
         [InlineData("GOT / HTTP/1.1", ' ', false, null)]
         [InlineData("ABC / HTTP/1.1", ' ', false, null)]
         [InlineData("PO / HTTP/1.1", ' ', false, null)]
         [InlineData("PO ST / HTTP/1.1", ' ', false, null)]
-        [InlineData("HTTP/1.0_\r", '\r', false, null)]
-        [InlineData("HTTP/1.1_\r", '\r', false, null)]
+        [InlineData("short ", ' ', false, null)]
+        public void GetsKnownMethode(string input, char endChar, bool expectedResult, string expectedKnownString)
```
Comment (cesarbs) [inline test/Microsoft.AspNet.Server.KestrelTests/MemoryPoolIterator2Tests.cs]: `Methode` -> `Method`

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 8]
```diff
@@ -68,6 +71,16 @@ private unsafe static long GetAsciiStringAsLong(string str)
                 return *(long*)ptr;
             }
         }
+        private unsafe static long GetMaskAsLong(byte[] bytes)
+        {
+            Debug.Assert(bytes.Length == 8, "Mask must be exactly 8 bytes long.");
+
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: Remove extra blank line.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 9]
```diff
@@ -26,35 +26,38 @@ public static class MemoryPoolIterator2Extensions
...
         private static long _http10VersionLong = GetAsciiStringAsLong("HTTP/1.0");
         private static long _http11VersionLong = GetAsciiStringAsLong("HTTP/1.1");
 
-        private const int PerfectHashDivisor = 37;
-        private static Tuple<long, string>[] _knownStrings = new Tuple<long, string>[PerfectHashDivisor];
+        private static long _maskFFFFFFFF = GetMaskAsLong(new byte[] { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff });
+        private static long _maskFFFFFFF0 = GetMaskAsLong(new byte[] { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00 });
+        private static long _maskFFFFFF00 = GetMaskAsLong(new byte[] { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, 0x00 });
+        private static long _maskFFFFF000 = GetMaskAsLong(new byte[] { 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, 0x00, 0x00 });
+        private static long _maskFFFF0000 = GetMaskAsLong(new byte[] { 0xff, 0xff, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00 });
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: Rename the masks to something more explanatory, e.g. `_mask4Chars`.

**Comment by cesarbs** [comment]
I did some tests of my own with your change and it went from 1,000,000 requests @ ~1200ms to ~900ms. Sweet :+1: 

There are a few more things I'd like you to address and then we'll be ready to merge.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 10]
```diff
@@ -262,54 +274,42 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
-                return false;
+                knownString = Http11Version;
+                return true;
             }
-
-            inputLong &= (long)(unchecked((ulong)~0) >> ((sizeof(long) - inputLength) * 8));
-
-            var value = _knownStrings[inputLong % PerfectHashDivisor];
-            if (value != null && value.Item1 == inputLong)
+            if (value == _http10VersionLong)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: `else if`

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 11]
```diff
@@ -262,54 +274,42 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
             {
-                knownString = value.Item2;
+                knownString = Http10Version;
+                return true;
             }
-
-            return knownString != null;
+            knownString = null;
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: This is redundant; remove.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 12]
```diff
@@ -262,54 +274,42 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
+            return false;
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: Add a blank line between `}` and `return`.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 13]
```diff
@@ -262,54 +274,42 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
+        }
 
-            var inputLong = begin.PeekLong();
-
-            if (inputLong == -1)
+        public static bool PeekKnownVersion(this MemoryPoolIterator2 begin, out string knownString)
+        {
+            knownString = null;
+            var value = begin.PeekLong();
+            if (value == _http11VersionLong)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: Add blank line before `if` block

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs [hunk 14]
```diff
@@ -90,6 +90,49 @@ public int Take()
             }
         }
 
+        public void Skip(int bytesToSkip)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: Please add unit tests for this new method. Have a look at `Microsoft.AspNet.Server.KestrelTests.MemoryPoolIterator2Tests.PeekLongAtBlockBoundary` if you need an idea as to how to test close to block boundaries.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 15]
```diff
@@ -262,54 +274,42 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
-        /// The number 37 was chosen because that number allows for a perfect hash of the set of
-        /// "known strings" (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE, HTTP/1.0 and HTTP/1.1, where strings
-        /// with less than 8 characters have 0s appended to their ends to fill for the missing bytes).
-        /// </remarks>
-        /// <param name="begin">The iterator from which to start the known string lookup.</param>
-        /// <param name="end">The iterator pointing to the end of the input string.</param>
-        /// <param name="knownString">A reference to a pre-allocated known string, if the input matches any.</param>
-        /// <returns><c>true</c> if the input matches a known string, <c>false</c> otherwise.</returns>
-        public static bool GetKnownString(this MemoryPoolIterator2 begin, MemoryPoolIterator2 end, out string knownString)
+        public static bool PeekKnownMethod(this MemoryPoolIterator2 begin, out string knownString)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: I'd rather keep the same method name as before (`Peek` -> `Get`).

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 16]
```diff
@@ -262,54 +274,42 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
+                    return true;
+                }
             }
+            return false;
+        }
 
-            var inputLong = begin.PeekLong();
-
-            if (inputLong == -1)
+        public static bool PeekKnownVersion(this MemoryPoolIterator2 begin, out string knownString)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: `Peek` -> `Get`

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 17]
```diff
@@ -262,54 +274,42 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
-        /// Since all of those fit in at most 8 bytes, they can be optimally looked up by reading those bytes as a long. Once
-        /// in that format, uninteresting bits are cleared and the remaining long modulo 37 is looked up in a table.
-        /// The number 37 was chosen because that number allows for a perfect hash of the set of
-        /// "known strings" (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE, HTTP/1.0 and HTTP/1.1, where strings
-        /// with less than 8 characters have 0s appended to their ends to fill for the missing bytes).
-        /// </remarks>
-        /// <param name="begin">The iterator from which to start the known string lookup.</param>
-        /// <param name="end">The iterator pointing to the end of the input string.</param>
-        /// <param name="knownString">A reference to a pre-allocated known string, if the input matches any.</param>
-        /// <returns><c>true</c> if the input matches a known string, <c>false</c> otherwise.</returns>
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: Please restore documentation with appropriate changes for the new logic.

**Comment by Anderman** [comment]
Thx for the comment. The aspnet team has the best repro's. Corefx has not that quality and coreclr is no fun at all. but maybe that's because my lack of knowlegde of c++. (I did a JIT change for an unsafe vector copy (-; and tried an intrinsic version of buffer.blockcopy )

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 18]
```diff
@@ -704,25 +704,24 @@ private void WriteSuffix()
             }
         }
 
-        protected bool TakeStartLine(SocketInput input)
+        public bool TakeStartLine(SocketInput input)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: Revert

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 19]
```diff
@@ -266,50 +278,81 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
         /// Checks that up to 8 bytes between <paramref name="begin"/> and <paramref name="end"/> correspond to a known HTTP string.
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: Doc references non-existing parameter (`end`).

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 20]
```diff
@@ -266,50 +278,81 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
         /// </summary>
         /// <remarks>
-        /// A "known HTTP string" can be an HTTP method name defined in the HTTP/1.1 RFC or an HTTP version (HTTP/1.0 or HTTP/1.1).
+        /// A "known HTTP method" can be an HTTP method name defined in the HTTP/1.1 RFC.
         /// Since all of those fit in at most 8 bytes, they can be optimally looked up by reading those bytes as a long. Once
-        /// in that format, uninteresting bits are cleared and the remaining long modulo 37 is looked up in a table.
-        /// The number 37 was chosen because that number allows for a perfect hash of the set of
-        /// "known strings" (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE, HTTP/1.0 and HTTP/1.1, where strings
-        /// with less than 8 characters have 0s appended to their ends to fill for the missing bytes).
+        /// in that format, it can be checked agains the known method.
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: typo: against

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 21]
```diff
@@ -266,50 +278,81 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
         /// <remarks>
-        /// A "known HTTP string" can be an HTTP method name defined in the HTTP/1.1 RFC or an HTTP version (HTTP/1.0 or HTTP/1.1).
+        /// A "known HTTP method" can be an HTTP method name defined in the HTTP/1.1 RFC.
         /// Since all of those fit in at most 8 bytes, they can be optimally looked up by reading those bytes as a long. Once
-        /// in that format, uninteresting bits are cleared and the remaining long modulo 37 is looked up in a table.
-        /// The number 37 was chosen because that number allows for a perfect hash of the set of
-        /// "known strings" (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE, HTTP/1.0 and HTTP/1.1, where strings
-        /// with less than 8 characters have 0s appended to their ends to fill for the missing bytes).
+        /// in that format, it can be checked agains the known method. 
+        /// The Known Methods (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE) are all less then 8 bytes
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: then -> than

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 22]
```diff
@@ -266,50 +278,81 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
-        /// A "known HTTP string" can be an HTTP method name defined in the HTTP/1.1 RFC or an HTTP version (HTTP/1.0 or HTTP/1.1).
+        /// A "known HTTP method" can be an HTTP method name defined in the HTTP/1.1 RFC.
         /// Since all of those fit in at most 8 bytes, they can be optimally looked up by reading those bytes as a long. Once
-        /// in that format, uninteresting bits are cleared and the remaining long modulo 37 is looked up in a table.
-        /// The number 37 was chosen because that number allows for a perfect hash of the set of
-        /// "known strings" (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE, HTTP/1.0 and HTTP/1.1, where strings
-        /// with less than 8 characters have 0s appended to their ends to fill for the missing bytes).
+        /// in that format, it can be checked agains the known method. 
+        /// The Known Methods (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE) are all less then 8 bytes 
+        /// and will be compared with the required space. A mask is used if the Known method is less then 8 bytes.
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: than -> than

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 23]
```diff
@@ -266,50 +278,81 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
+        /// To optimize performance the GET method will be checked first
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: end sentence with a period

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 24]
```diff
@@ -266,50 +278,81 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
         /// </remarks>
         /// <param name="begin">The iterator from which to start the known string lookup.</param>
-        /// <param name="end">The iterator pointing to the end of the input string.</param>
-        /// <param name="knownString">A reference to a pre-allocated known string, if the input matches any.</param>
+        /// <param name="scan">If we found a valid method, the scan will be updated to new position</param>
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: I think you meant "theN scan" instead of "the scan"?

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 25]
```diff
@@ -266,50 +278,81 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
-            var inputLong = begin.PeekLong();
+            return false;
+        }
 
-            if (inputLong == -1)
+        /// <summary>
+        /// Checks that up to 8 bytes between <paramref name="begin"/> and <paramref name="end"/> correspond to a known HTTP string.
+        /// </summary>
+        /// <remarks>
+        /// A "known HTTP version" Is a version defined in the HTTP/1.1 (HTTP/1.0 or HTTP/1.1).
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: I'd replace this sentence with `A "known HTTP version" is either HTTP/1.0 or HTTP/1.1.`

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 26]
```diff
@@ -266,50 +278,81 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
+        /// Since those fit in 8 bytes, they can be optimally looked up by reading those bytes as a long. Once
+        /// in that format, it can be checked agains the known versions.
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: against

**Comment by cesarbs** [comment]
A few more comments. You'll need to rebase since there are conflicts with the dev branch.

**Comment by Anderman** [comment]
@cesarbs I accidentally-closed this pull request. I don't know how to reopen this PR

**Comment by cesarbs** [comment]
Reopened.
